codex-rs/app-server/src/bespoke_event_handling.rs

```diff
@@ -38,14 +43,58 @@ use codex_core::protocol::McpToolCallBeginEvent;
 use codex_core::protocol::McpToolCallEndEvent;
 use codex_core::protocol::Op;
 use codex_core::protocol::ReviewDecision;
+use codex_core::protocol::TokenUsage;
 use codex_protocol::ConversationId;
+use std::collections::HashMap;
 use std::convert::TryFrom;
 use std::sync::Arc;
+use std::sync::OnceLock;
+use tokio::sync::Mutex;
 use tokio::sync::oneshot;
 use tracing::error;
 
 type JsonValue = serde_json::Value;
 
+#[derive(Default, Clone)]
+struct TurnAccum {
+    last_total_token_usage: Option<TokenUsage>,
+    last_error_message: Option<String>,
+}
+
+type TurnKey = (ConversationId, String);
+
+static TURN_STATE: OnceLock<Arc<Mutex<HashMap<TurnKey, TurnAccum>>>> = OnceLock::new();
+
+fn turn_state() -> &'static Arc<Mutex<HashMap<TurnKey, TurnAccum>>> {
+    TURN_STATE.get_or_init(|| Arc::new(Mutex::new(HashMap::new())))
+}
+
+async fn take_turn_accum(
+    conversation_id: ConversationId,
+    event_id: &str,
+) -> (Option<TokenUsage>, Option<String>) {
+    let key = (conversation_id, event_id.to_string());
+    let state = turn_state();
+    let mut map = state.lock().await;
+    let entry = map.remove(&key).unwrap_or_default();
+    (entry.last_total_token_usage, entry.last_error_message)
```

Comment:
Drop the lock as soon as you no longer need it:

```suggestion
    let entry = {
        let mut map = state.lock().await;
        map.remove(&key).unwrap_or_default()
    };
    (entry.last_total_token_usage, entry.last_error_message)
```